src/hybrid/hybrid_search_result.c

```diff
@@ -0,0 +1,164 @@
+ */
+void HybridSearchResult_Free(HybridSearchResult* result) {
+  if (!result) {
+    return;
+  }
+
+  // Free individual SearchResults with array_free_ex
+  array_free_ex(result->searchResults, {
+    SearchResult **sr = (SearchResult**)ptr;
+    if (*sr) {
+      SearchResult_Destroy(*sr);
+      rm_free(*sr);
+    }
+  });
+
+  array_free(result->hasResults);
+
+  rm_free(result);
+}
+
+/**
+ * Store a SearchResult from a source into the HybridSearchResult.
+ * Updates the score of the SearchResult and marks the source as having results.
+ */
+void HybridSearchResult_StoreResult(HybridSearchResult* hybridResult, SearchResult* searchResult, int sourceIndex) {
+  if (!hybridResult || !searchResult ) {
+    return;
+  }
+  RS_ASSERT(sourceIndex < hybridResult->numSources);
+
+  // Store the SearchResult from this source (preserving all data)
+  hybridResult->searchResults[sourceIndex] = searchResult;
+  hybridResult->hasResults[sourceIndex] = true;
+}
+
+
+/**
+ * Apply hybrid scoring to compute combined score from multiple sources.
+ * Supports both RRF (with ranks) and Linear (with scores) hybrid scoring.
+ */
+double ApplyHybridScoring(HybridSearchResult *hybridResult, int8_t targetIndex, HybridScoringContext *scoringCtx) {
+  if (!hybridResult || !scoringCtx) {
+    return 0.0;
+  }
+  RS_ASSERT(hybridResult->hasResults[targetIndex]);
+
+  // Extract values from SearchResults
+  arrayof(double) values = array_newlen(double, hybridResult->numSources);
+  for (size_t i = 0; i < hybridResult->numSources; i++) {
+    if (hybridResult->hasResults[i] && hybridResult->searchResults[i]) {
+      // Note: SearchResult->score contains ranks for RRF, scores for Linear
+      // This is set correctly by upstream processors based on scoring type
+      values[i] = hybridResult->searchResults[i]->score;
+    } else {
+      values[i] = 0.0;  // Default value for missing results
+    }
+  }
+
+  // Calculate hybrid score using generic scoring function
+  HybridScoringFunction scoringFunc = GetScoringFunction(scoringCtx->scoringType);
+  double result = scoringFunc(scoringCtx, values, hybridResult->hasResults, hybridResult->numSources);
+
+  array_free(values);
+  return result;
+}
+
+/**
+ * Main function to merge SearchResults from multiple upstreams into a single comprehensive result.
+ * Finds the primary result, computes hybrid score, merges flags, and returns the merged result.
+ * This function transfers ownership of the primary result from the HybridSearchResult to the caller.
+ */
+SearchResult* MergeSearchResults(HybridSearchResult *hybridResult, HybridScoringContext *scoringCtx) {
+  RS_ASSERT(hybridResult && scoringCtx);
+
+  // Find the primary result (first non-null result)
+  SearchResult *primary = NULL;
+  int8_t targetIndex = -1;
+  for (size_t i = 0; i < hybridResult->numSources; i++) {
+    if (hybridResult->hasResults[i] && hybridResult->searchResults[i]) {
+      primary = hybridResult->searchResults[i];
```

Comment:
Let add an assertion for `primary`

Reply:
I mean an assertion for `primary` - `RS_ASSERT(primary)`
And removing the second condition in the `if` - that part was less clear